File: neetcode/Array/Top_K_Elements/TopKElements.java
// https://leetcode.com/problems/top-k-frequent-elements
// https://neetcode.io/problems/top-k-elements-in-list?list=blind75

class TopKElements {
    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer, Integer> map = new HashMap<>();
 
        for (int num : nums) {
            if ( !map.containsKey(num) ) {
                map.put(num, 1);
            } else {
                map.put(num, map.get(num) + 1);
            }
        }

        List<int []> arr = new ArrayList<>();
        for ( Map.Entry<Integer, Integer> entry : map.entrySet() ) {
            arr.add( new int[]{ entry.getValue(), entry.getKey() });
        }
        arr.sort((a,b) -> b[0] - a[0]);

        int[] arr1 = new int[k];
        for ( int i=0; i<k; i++ ) {
            arr1[i] = arr.get(i)[1];
        }

        return arr1;
    }
}

File: neetcode/python/Array/top_k_elements.py

# https://leetcode.com/problems/top-k-frequent-elements
# https://neetcode.io/problems/top-k-elements-in-list?list=blind75

# O(n logn) time complexity where n is the numbers in the list
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        count = {}
        for num in nums:
            count[num] = count.get(num, 0) + 1
        
        arr = []
        for num in count.keys():
            arr.append([count[num], num])
        arr.sort()

        res = []
        for i in range(k):
            res.append(arr.pop()[1])
        
        return res


# O(n logk) time complexity where n is the numbers in the list
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        count = {}
        for num in nums:
            count[num] = count.get(num, 0) + 1
        
        heap = []
        for num in count.keys():
            heapq.heappush(heap, [count[num], num])
            if (len(heap) > k):
                heapq.heappop(heap)

        res = []
        for i in range(k):
            res.append(heapq.heappop(heap)[1])
        
        return res
